src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -3941,3 +4086,363 @@ CompareInsertValuesByShardId(const void *leftElement, const void *rightElement)
+ * see header of routine ErrorIfDistTablesNotColocated for details, raises an exception
+ * for all other combinations.
+ */
+static DeferredErrorMessage *
+ErrorIfMergeHasUnsupportedTables(Query *parse, List *rangeTableList,
+								 PlannerRestrictionContext *restrictionContext)
+{
+	List *distTablesList = NIL;
+	bool foundLocalTables = false;
+
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		Oid relationId = rangeTableEntry->relid;
+
+		switch (rangeTableEntry->rtekind)
+		{
+			case RTE_RELATION:
+			{
+				/* Check the relation type */
+				break;
+			}
+
+			case RTE_SUBQUERY:
+			case RTE_FUNCTION:
+			case RTE_TABLEFUNC:
+			case RTE_VALUES:
+			case RTE_JOIN:
+			case RTE_CTE:
+			{
+				/* Skip them as base table(s) will be checked */
+				continue;
+			}
+
+			/*
+			 * RTE_NAMEDTUPLESTORE is typically used in ephmeral named relations,
+			 * such as, trigger data; until we find a genuine use case, raise an
+			 * exception.
+			 * RTE_RESULT is a node added by the planner and we shouldn't
+			 * encounter it in the parse tree.
+			 */
+			case RTE_NAMEDTUPLESTORE:
+			case RTE_RESULT:
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "MERGE command is not supported with "
+									 "Tuplestores and results",
+									 NULL, NULL);
+			}
+
+			default:
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "MERGE command: Unrecognized range table entry.",
+									 NULL, NULL);
+			}
+		}
+
+		/* RTE Relation can be of various types, check them now */
+
+		/* skip the regular views as they are replaced with subqueries */
+		if (rangeTableEntry->relkind == RELKIND_VIEW)
+		{
+			continue;
+		}
+
+		if (rangeTableEntry->relkind == RELKIND_MATVIEW ||
+			rangeTableEntry->relkind == RELKIND_FOREIGN_TABLE)
+		{
+			/* Materialized view or Foreign table as target is not allowed */
+			if (IsMergeAllowedOnRelation(parse, rangeTableEntry))
+			{
+				/* Non target relation is ok */
+				continue;
+			}
+			else
+			{
+				/* Usually we don't reach this exception as the Postgres parser catches it */
+				StringInfo errorMessage = makeStringInfo();
+				appendStringInfo(errorMessage,
+								 "MERGE command is not allowed on "
+								 "relation type(relkind:%c)", rangeTableEntry->relkind);
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED, errorMessage->data,
+									 NULL, NULL);
+			}
+		}
+
+		if (rangeTableEntry->relkind != RELKIND_RELATION &&
+			rangeTableEntry->relkind != RELKIND_PARTITIONED_TABLE)
+		{
+			StringInfo errorMessage = makeStringInfo();
+			appendStringInfo(errorMessage, "Unexpected table type(relkind:%c) "
+										   "in MERGE command", rangeTableEntry->relkind);
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED, errorMessage->data,
+								 NULL, NULL);
+		}
+
+		Assert(rangeTableEntry->relid != 0);
+
+		/* Reference tables are not supported yet */
+		if (IsCitusTableType(relationId, REFERENCE_TABLE))
```

Comment:
I think supporting reference table as a target is not easy.

But, what is the problem of having reference table in the "source" joined with a distributed tables / subqueries?

(again, I'm fine to omit in this patch, but along with recursive planning, this seems like a limitation that could hurt users)

Reply:
>>But, what is the problem of having reference table in the "source" joined with a distributed tables / subqueries?

We have to manage "NOT MATCHED" rows to avoid duplicate INSERTs, but I guess this problem of OUTER join with Reference table as outer table is recently solved/supported. It's doable for MERGE too, may be with some tweaks?

Marco and I briefly talked about it during our prior review meeting and decided to support reference tables in its own separate patch. As we have legroom for 11.3, I believe we can shoot for reference tables too in our next phase.

Reply:
Alright, I have been reading some PG code, and wanted to summarize my current understanding regarding reference table / recursive planning support. Let's discuss if this makes sense to you as well.

PG [converts](https://github.com/postgres/postgres/blob/eae0e20deffb0a73f7cb0e94746f94a1347e71b1/src/backend/optimizer/prep/prepjointree.c#L152) the merge command into `... target RIGHT JOIN source ...`. In that case, `source` is in the outer part of an outer join. Hence, we cannot allow `source` to be [`recurring tuple`](https://github.com/citusdata/citus/blob/1e415bb1c306259cf30dcaf2e3e8e32dda9701de/src/backend/distributed/planner/query_pushdown_planning.c#L50-L64) and `target` to be a distributed table -- and push down the command. If we'd allow, then `WHEN NOT MATCHED` would yield duplicate results on the `target`. 

- In the first iteration of Phase-3, we can follow what this PR does. Strictly require all tables to be colocated
- In the second iteration of Phase-3, we can relax the checks a bit further.
    -  The source could involve reference tables 
    - We let `source` handled by recursive planning. We still cannot allow repartition joins / non-colocated tables after recursive planning. In other words, for all the remaining distributed tables in the `source`, we should enforce what the Phase-1 enforces (`IsPartitionColumnInMergeSource` and `ErrorIfTablesNotColocatedAndJoinedOnDistColumn`).

- In the third iteration of Phase-3 (not sure we should/can do in 11.3), if the above would not yield a successful plan (probably only when `!IsPartitionColumnInMergeSource()`), we can plan and execute `source` subquery, then repartition the results based on `target` table's distribution key similar to `INSERT_SELECT_REPARTITION`.
    - This would probably also allow merge into `reference table` case  

 

> It's doable for MERGE too, may be with some tweaks?

Probably not, because at this point, the `target` relation would be recursively planned according to what we support today via recursive planning. And, it (probably?) doesn't make sense to recursively plan target table anyway.

> Marco and I briefly talked about it during our prior review meeting and decided to support reference tables in its own separate patch. As we have legroom for 11.3, I believe we can shoot for reference tables too in our next phase.

Sounds good, lets focus on this PR first, which is going to be the backbone anyway. Reference tables in the `source` vs `target` are probably different problems, and in the `source` is probably more useful.

(ps: if all tables are reference tables, we could pushdown as-is, but lets ignore that in this PR as well)

Reply:
Agree, let's keep reference table aside for now, and address it in a different PR.